### PR TITLE
Fix retrying saving anomaly results

### PR DESCRIPTION
**Pull request by kaituo**
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/153

*Description of changes:*
Previously, AD throws an exception when retrying saving an anomaly result, indicating autoGeneratedTimestamp should not be set. We can see the exception because we reuse the same IndexRequest, where autoGeneratedTimestamp is already set. This PR creates a new IndexRequest by copying the original request without autoGeneratedTimestamp.

This PR also adds detector id to AnomalyDetectionException’s toString method so that we can see the id in the log. This PR also adds more logging to SearchFeatureDao.

Testing done:
* Manually test to make sure the exception is gone after my fix.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyResultHandler.java [hunk 1]
```diff
@@ -192,7 +192,12 @@ void saveDetectorResult(IndexRequest indexRequest, String context, Iterator<Time
             } else {
                 TimeValue nextDelay = backoff.next();
                 LOG.warn(RETRY_SAVING_ERR_MSG + context, cause);
-                threadPool.schedule(() -> saveDetectorResult(indexRequest, context, backoff), nextDelay, ThreadPool.Names.SAME);
+                // copy original request's source without other information like autoGeneratedTimestamp
+                // otherwise, an exception will be thrown indicating autoGeneratedTimestamp should not be set
+                // while request id is already set (id is set because we have already sent the request before).
+                IndexRequest newReuqest = new IndexRequest(AnomalyResult.ANOMALY_RESULT_INDEX);
```
Comment (wnbts) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyResultHandler.java:198]: Minor. Should the index name be from the original index request?

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/AnomalyResultHandler.java:198]: This is AnomalyResultHandler, so its index should be AnomalyResult.ANOMALY_RESULT_INDEX.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/SearchFeatureDao.java [hunk 2]
```diff
@@ -367,6 +380,7 @@ public void getFeaturesForSampledPeriods(
         ActionListener<Optional<Entry<double[][], Integer>>> listener
     ) {
         Map<Long, double[]> cache = new HashMap<>();
+        logger.info(String.format("Getting features for detector %s starting %d", detector.getDetectorId(), endTime));
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/SearchFeatureDao.java:383]: `starting %d` -> `ending at %d`?

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/SearchFeatureDao.java:383]: Good catch.  we don't know the start time at this point.  Changed to "logger.info(String.format("Getting features for detector %s ending at %d", detector.getDetectorId(), endTime));"

**Pull request merged**